[WIP] Tpetra: new constructor with 4 maps and local matrix

aprokop opened the pull request:
Fix #1536.

Remaining items:
- [x] Add a unit test
- [x] ~Figure out if it can be done without code duplication in `CrsMatrix`, similarly to `CrsGraph`.~
- [ ] Squash and clean up

mhoemmen commented:
If you don't intend for the graph / matrix to be fill complete when the constructor returns, this may take very little work.  The 0/1-argument version of `CrsGraph::fillComplete` already does what you want: it uses the graph's existing domain and range Maps if they are not null.  The 0/1-argument version of `CrsMatrix::fillComplete` would need to be changed to imitate this behavior; right now it only uses the graph's existing domain and range Maps in the case where the matrix's graph is const (`isStaticGraph()`) and fill complete.

aprokop commented:
@mhoemmen I think this constructor has to fill complete for it to be consistent with the 2 maps+local matrix constructor.

aprokop commented:
@mhoemmen I stuck a test for 4 maps with a local matrix in a triangular test, The test actually tests that a newly constructed matrix does a proper `apply`, and thus this should sufficiently stress the import/export and therefore maps.

mhoemmen commented on packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp:
```diff
@@ -660,6 +660,19 @@ namespace Tpetra {
+            const Teuchos::RCP<const map_type>& rangeMap,
+            const local_graph_type& k_local_graph_,
+            const Teuchos::RCP<Teuchos::ParameterList>& params)
+    : CrsGraph<LocalOrdinal, GlobalOrdinal, Node, classic>(rowMap, colMap, k_local_graph_, params)
```
The CrsGraph constructor to which you delegate here will (expensively) create an Import object, which you'll throw away below when calling makeImportExport again.  It might be better to factor out common code in both constructors, then implement them separately.

mhoemmen commented on packages/tpetra/core/test/CrsMatrix/CrsMatrix_UnitTests2.cpp:
```diff
@@ -430,6 +431,20 @@ inline void tupleToArray(Array<T> &arr, const tuple &tup)
+    // test the constructors based on 4 maps + local matri_crsx
+    RCP<MAT> tri_crs_2 = rcp(new MAT(tri_crs->getRowMap(), tri_crs->getColMap(), tri_crs->getDomainMap(),
+            tri_crs->getRangeMap(), tri_crs->getLocalMatrix()) );
+    // test the action
```
Should also test whether the resulting matrix is fill complete, and whether the Maps and Import (and Export, if applicable) objects match.

aprokop commented:
Addressed the comments. Ideally, I would actually prefer the interface to be like this:
```c++
CrsGraph::CrsGraph(local_graph, map, map, map=null, map=null, params=null)
```
This would remove code duplication. But it's not backward compatible.

mhoemmen commented:
@aprokop wrote:
> This would remove code duplication. But it's not backward compatible.

Would that conflict with any existing constructors?  If not, then it's perfectly backwards compatible.  You could make the existing (map, map, local_graph) constructor just delegate to the new one.

mhoemmen commented on packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp:
```diff
@@ -2151,6 +2182,10 @@ namespace Tpetra {
 
+    /// \brief Forces computation of local triangular properties if they have
+    /// not been computed yet.
+    void computeTriangularProperties();
```
I would call this "computeLocalTrangularProperties" just to make clear that it's local.

mhoemmen replied:
Also, please mark this with lots of warning documentation saying "THIS IS FOR EXPERTS ONLY" and "NO PROMISE OF BACKWARDS COMPATIBILITY; THIS METHOD MAY CHANGE OR DISAPPEAR AT ANY TIME."

aprokop replied:
Will do both.

mhoemmen commented on packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp:
```diff
@@ -4735,6 +4752,70 @@ namespace Tpetra {
+  template <class LocalOrdinal, class GlobalOrdinal, class Node, const bool classic>
+  void
+  CrsGraph<LocalOrdinal, GlobalOrdinal, Node, classic>::
+  computeTriangularProperties ()
```
See above note (on calling this "computeLocalTriangularProperties").  Also, it might make sense for this to be a free function that takes a `local_graph_type` input, instead of an instance method of (the global data structure) `Tpetra::CrsGraph`.  That's because the function only knows how to work on the `local_graph_type` instance, not on any of CrsGraph's other data structure options.

aprokop replied:
OK.

aprokop replied:
I decided not to make it a free function as it still sets some CrsGraph members.

mhoemmen commented on packages/tpetra/core/test/CrsMatrix/CrsMatrix_UnitTests2.cpp:
```diff
@@ -430,6 +431,28 @@ inline void tupleToArray(Array<T> &arr, const tuple &tup)
+    auto importer = tri_crs_2->getGraph()->getImporter();
+    TEST_EQUALITY(tri_crs->getDomainMap()->isSameAs(*(importer->getSourceMap())), true);
+    TEST_EQUALITY(tri_crs->getColMap()   ->isSameAs(*(importer->getTargetMap())), true);
+    TEST_EQUALITY(tri_crs->getRowMap()   ->isSameAs(*(exporter->getSourceMap())), true);
```
Is the Export object trivial (and therefore null) here?  You may want to protect these equality tests with `if (! importer.is_null ())` resp. `if (! exporter.is_null ())`.

aprokop replied:
In this example, domain map != column map and row map != range map, so this works.

aprokop replied:
Though, you may be right, and it breaks the serial build. I'll add it just in case.

mhoemmen commented on packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp:
```diff
@@ -4735,6 +4752,70 @@ namespace Tpetra {
+
+    // Compute triangular properties
+    const size_t numLocalRows = getNodeNumRows ();
+    for (size_t localRow = 0; localRow < numLocalRows; ++localRow) {
```
Why not wrap this in a `Kokkos::parallel_reduce` using a `KOKKOS_LAMBDA`?

mhoemmen replied:
It might be a bit more complicated -- you would need to define a struct for the reduction results.

aprokop replied:
I think this is beyond the scope of this PR, this code was lifted from the original constructor.

aprokop commented:
@mhoemmen I updated the PR. From my perspective, it's complete. It is not addressing the parallel region issue and the issue of changing API. If you want me to, I could create new PRs for those.

mhoemmen commented:
@aprokop wrote:
> It is not addressing the parallel region issue and the issue of changing API. If you want me to, I could create new PRs for those.

I'm not worried about the parallel region issue for now.  We can defer that for later.

Regarding the API change: We shouldn't add public constructors unless they have the interface we ultimately want.  For now, would it make sense to make the constructor look like this: (local data structure, Map, Map, Map, Map)?

aprokop commented:
@mhoemmen I updated the API to what we want. The only remaining question is whether the ParameterList should be provided. It was in the original constructor, but never used. But it may benefit to leave it there for the future.

mhoemmen commented:
@aprokop wrote:
> The only remaining question is whether the ParameterList should be provided. It was in the original constructor, but never used. But it may benefit to leave it there for the future.

Please do leave it there; thanks!

aprokop commented:
@mhoemmen Then I think this PR is done. Once you approve, I'll clean it up, and it's ready for merge. I'm still reinstalling my spack modules after Fedora upgrade, so may take a bit of time to run the checkin script.

mhoemmen commented on packages/tpetra/core/src/Tpetra_CrsGraph_decl.hpp:
```diff
@@ -539,6 +539,37 @@ namespace Tpetra {
+    /// \param params [in/out] Optional list of parameters.  If not
+    ///   null, any missing parameters will be filled in with their
+    ///   default values.
+    CrsGraph (const local_graph_type& lclGraph,
```
Thanks for moving `local_graph_type` to the first argument :-) .

mhoemmen commented:
@aprokop wrote:
> Once you approve, I'll clean it up, and it's ready for merge. I'm still reinstalling my spack modules after Fedora upgrade, so may take a bit of time to run the checkin script.

If you're able to push yourself, please feel free to do so after testing.  Otherwise, just let me know once tests have passed and I'll approve the PR.  Thanks!

aprokop commented:
Did the checkin push: 86b71bfb2a08737caef724561818baf123e097c5.

aprokop commented:
@mhoemmen Thanks a lot for very useful feedback!

mhoemmen commented:
@aprokop Glad to help, and thanks for working on this and being open to changes!

aprokop commented:
Oh, the moment you realize that you forgot an API change in a month old PR...
The API for the `CrsGraph` was changed to `(local_matrix, map, map, map, map)`, but the API for the `CrsMatrix` is still `(map, map, map, map, local_matrix)`. I'll do another PR, and hope that nobody used the interface in the past month.

mhoemmen commented:
@aprokop wrote:
> I'll do another PR, and hope that nobody used the interface in the past month.

OK since we haven't done a minor release this month :-)
